src/rust/lib/graph-editor/src/component/edge.rs

```diff
@@ -44,23 +47,280 @@ const NODE_PADDING       : f32 = node::SHADOW_SIZE;
 const PADDING            : f32 = 4.0;
 const RIGHT_ANGLE        : f32 = std::f32::consts::PI / 2.0;
 
+const SNAP_THRESHOLD     : f32 = 1.2 * LINE_SHAPE_WIDTH;
+
 const INFINITE : f32 = 99999.0;
 
 
 
+// ========================
+// === Edge Shape Trait ===
+// ========================
+
+/// Edge shape defines the common behaviour of the sub-shapes used to create a Edge.
+trait EdgeShape: ensogl::display::Object {
+    fn set_hover_split_offset(&self, offset:Vector2<f32>);
+    fn set_hover_split_rotation(&self, angle:f32);
+    /// Snaps the coordinates given in the shape local coordinate system to the shape and returns
+    /// the shape local coordinates. If no snapping is possible, returns `None`.
+    fn snap_to_shape_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>>;
+    fn sprite(&self) -> &Sprite;
+
+    /// Return the angle perpendicular to the shape at the given point. Defaults to zero, if not
+    /// implemented.
+    fn cut_angle_for_point(&self, point:Vector2<f32>) -> nalgebra::Rotation2<f32>  {
+        let local = self.to_shape_coordinate_system(point);
+        self.cut_angle_for_point_local(local)
+    }
+
+    fn cut_angle_for_point_local(&self, _point:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+        nalgebra::Rotation2::new(0.0)
+    }
+
+    /// Snaps the coordinates given in the global coordinate system to the shape and returns
+    /// the global coordinates. If no snapping is possible, returns `None`.
+    fn snap_to_shape(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+        let local          = self.to_shape_coordinate_system(point);
+        let local_snapped  = self.snap_to_shape_local(local)?;
+        let global_snapped = self.from_shape_to_global_coordinate_system(local_snapped);
+        Some(global_snapped)
+    }
+
+    /// Set the hover split for this shape. The `split_data` indicates where the shape should be
+    /// split and which side should be recolored.
+    fn enable_hover_split(&self, split_data:SplitData) {
+        let SplitData {split_position,area,cut_angle} = split_data;
+        // Compute rotation in shape local coordinate system.
+        let base_rotation        = self.display_object().rotation().z;
+        let hover_split_rotation = cut_angle + base_rotation;
+        match area {
+            Area::Above => self.set_hover_split_rotation(hover_split_rotation),
+            Area::Below => {
+                self.set_hover_split_rotation(hover_split_rotation + 2.0 * RIGHT_ANGLE)
+            },
+        }
+        // Compute position in shape local coordinate system.
+        let offset = self.to_shape_coordinate_system(split_position);
+        self.set_hover_split_offset(offset)
+    }
+
+    /// Disable the hover split on this shape.
+    fn disable_hover_split(&self) {
+        self.set_hover_split_offset(Vector2::new(INFINITE, INFINITE));
```

Comment:
fyi, since few weeks, we have exported functions for easier construction of vectors. You can do now this: `Vector2(INFINITE,INFINITE)` instead of `Vector2::new(INFINITE,INFINITE)`

Reply:
This looks like introducing other languages constructors syntax. Is this not violating some of our own style guidelines, too? (Uppercase functions?).